Constructions/eps_additon.ml

```diff
@@ -0,0 +1,2136 @@
+  ==> (\b0:string. (eval (add_ebin (App a0 a1) (y:epsilon)) to (nat))) (w:string) =
+  add_unary ((\b0:string. (eval (App a0 a1) to (nat))) (w:string))
+  ((\b0:string. (eval (y:epsilon) to (nat))) (w:string))) =
+  (proper_nat_construct (App a0 a1) ==> proper_nat_construct (y:epsilon)
+  ==> (eval (add_ebin (App a0 a1) (y:epsilon)) to (nat)) =
+  add_unary (eval (App a0 a1) to (nat)) (eval (y:epsilon) to (nat)))`,
+  REWRITE_TAC[taut_lemma;IMP_CONJ] THEN 
+  REPEAT DISCH_TAC THEN
+  MP_TAC(SPECL [`(App a0 a1):epsilon`;`y:epsilon`] lemma6) THEN 
+  ASM_REWRITE_TAC[] THEN 
+  DISCH_TAC THEN 
+  MP_TAC(SPEC `(App a0 a1):epsilon` beta_red3) THEN 
+  MP_TAC(SPEC `y:epsilon` beta_red3) THEN 
+  MP_TAC(SPEC `(add_ebin (App a0 a1) y):epsilon` beta_red3) THEN
+  PROPER_TYPE_TAC(`(App a0 a1):epsilon`) THEN 
+  PROPER_TYPE_TAC(`y:epsilon`) THEN 
+  PROPER_TYPE_TAC(`(add_ebin (App a0 a1) y):epsilon`) THEN 
+  (PROPER_NOT_FREE_TAC `y:epsilon` `"j"` `TyBase "epsilon"`) THEN 
+  (PROPER_NOT_FREE_TAC `(App a0 a1):epsilon` `"j"` `TyBase "epsilon"`) THEN 
+  (PROPER_NOT_FREE_TAC `(add_ebin (App a0 a1) y):epsilon` `"j"` `TyBase "epsilon"`) THEN 
+  NAT_BETA_EVAL_RED(`(\j:epsilon. (eval (j) to (nat))) (y:epsilon)`) THEN 
+  NAT_BETA_EVAL_RED(`(\j:epsilon. (eval (j) to (nat))) ((App a0 a1):epsilon)`) THEN 
+  NAT_BETA_EVAL_RED(`(\j:epsilon. (eval (j) to (nat))) ((add_ebin (App a0 a1) y):epsilon)`) THEN 
+  REPEAT DISCH_TAC THEN 
+  ASM_REWRITE_TAC[]
+  );;
+
+let fin_beta_red8 = prove(
+  `(proper_nat_construct (App a0 a1) ==> proper_nat_construct y
+  ==> (\b1. (eval (add_ebin (App a0 a1) y) to (nat))) (w:type) =
+  add_unary ((\b1. (eval (App a0 a1) to (nat))) w)
+  ((\b1. (eval (y) to (nat))) w)) =
+  (proper_nat_construct (App a0 a1) ==> proper_nat_construct y
+  ==> (eval (add_ebin (App a0 a1) y) to (nat)) =
+  add_unary (eval (App a0 a1) to (nat)) (eval (y) to (nat)))`,
+  REWRITE_TAC[taut_lemma;IMP_CONJ] THEN 
+  REPEAT DISCH_TAC THEN
+  MP_TAC(SPECL [`(App a0 a1):epsilon`;`y:epsilon`] lemma6) THEN 
+  ASM_REWRITE_TAC[] THEN 
+  DISCH_TAC THEN 
+  MP_TAC(SPEC `(App a0 a1):epsilon` beta_red4) THEN 
+  MP_TAC(SPEC `y:epsilon` beta_red4) THEN 
+  MP_TAC(SPEC `(add_ebin (App a0 a1) y):epsilon` beta_red4) THEN
+  PROPER_TYPE_TAC(`(App a0 a1):epsilon`) THEN 
+  PROPER_TYPE_TAC(`y:epsilon`) THEN 
+  PROPER_TYPE_TAC(`(add_ebin (App a0 a1) y):epsilon`) THEN 
+  (PROPER_NOT_FREE_TAC `y:epsilon` `"j"` `TyBase "epsilon"`) THEN 
+  (PROPER_NOT_FREE_TAC `(App a0 a1):epsilon` `"j"` `TyBase "epsilon"`) THEN 
+  (PROPER_NOT_FREE_TAC `(add_ebin (App a0 a1) y):epsilon` `"j"` `TyBase "epsilon"`) THEN 
+  NAT_BETA_EVAL_RED(`(\j:epsilon. (eval (j) to (nat))) (y:epsilon)`) THEN 
+  NAT_BETA_EVAL_RED(`(\j:epsilon. (eval (j) to (nat))) ((App a0 a1):epsilon)`) THEN 
+  NAT_BETA_EVAL_RED(`(\j:epsilon. (eval (j) to (nat))) ((add_ebin (App a0 a1) y):epsilon)`) THEN 
+  REPEAT DISCH_TAC THEN 
+  ASM_REWRITE_TAC[]
+  );;
+
+let fin_beta_red9 = prove( 
```

Comment:
probably really worth analyzing the commonalities of these 9 `beta_red` proofs to try to pull out re-usable lemmas!